File: backend/src/main/java/cn/etarch/mao/harness/tool/SearchScope.java
package cn.etarch.mao.harness.tool;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;

/**
 * Normalizes a resolved search path (directory or file) into rg-compatible cwd + target.
 */
public final class SearchScope {

    private final Path cwd;
    private final String rgTarget;
    private final Path singleFile;

    private SearchScope(Path cwd, String rgTarget, Path singleFile) {
        this.cwd = cwd;
        this.rgTarget = rgTarget;
        this.singleFile = singleFile;
    }

    public static SearchScope from(Path resolved) throws IOException {
        if (Files.isRegularFile(resolved)) {
            Path parent = resolved.getParent();
            if (parent == null) {
                throw new IOException("Cannot search in file without parent directory: " + resolved);
            }
            return new SearchScope(parent, parent.relativize(resolved).toString(), resolved);
        }
        if (Files.isDirectory(resolved)) {
            return new SearchScope(resolved, ".", null);
        }
        throw new IOException("Search path does not exist or is not accessible: " + resolved);
    }

    public Path cwd() {
        return cwd;
    }

    public String rgTarget() {
        return rgTarget;
    }

    public boolean isSingleFile() {
        return singleFile != null;
    }

    public Path singleFile() {
        return singleFile;
    }

    public String outputFilePath(String rgFilePath, Path workspaceRoot) {
        if (singleFile != null) {
            Path normalized = singleFile.normalize();
            if (workspaceRoot != null && normalized.startsWith(workspaceRoot.normalize())) {
                return workspaceRoot.relativize(normalized).toString();
            }
            return normalized.getFileName().toString();
        }
        return relativizeRgPath(rgFilePath, cwd);
    }

    public static String relativizeRgPath(String pathStr, Path searchRoot) {
        String trimmed = pathStr.startsWith("./") ? pathStr.substring(2) : pathStr;
        Path path = Path.of(trimmed);
        if (path.isAbsolute()) {
            Path normalized = path.normalize();
            return normalized.startsWith(searchRoot)
                    ? searchRoot.relativize(normalized).toString()
                    : normalized.toString();
        }
        return path.normalize().toString();
    }
}

File: backend-ts/src/harness/tool/search-scope.ts
import { existsSync, statSync } from 'node:fs';
import path from 'node:path';

export class SearchScope {
  private constructor(
    readonly cwd: string,
    readonly rgTarget: string,
    readonly singleFile: string | null,
  ) {}

  static from(resolved: string): SearchScope {
    const stat = existsSync(resolved) ? statSync(resolved) : null;
    if (stat?.isFile()) {
      const parent = path.dirname(resolved);
      if (!parent) {
        throw new Error('Cannot search in file without parent directory: ' + resolved);
      }
      return new SearchScope(parent, path.relative(parent, resolved), resolved);
    }
    if (stat?.isDirectory()) {
      return new SearchScope(resolved, '.', null);
    }
    throw new Error('Search path does not exist or is not accessible: ' + resolved);
  }

  isSingleFile(): boolean {
    return this.singleFile != null;
  }

  outputFilePath(rgFilePath: string, workspaceRoot: string | null): string {
    if (this.singleFile != null) {
      const normalized = path.resolve(this.singleFile);
      if (workspaceRoot != null) {
        const root = path.resolve(workspaceRoot);
        const rel = path.relative(root, normalized);
        if (!rel.startsWith('..') && !path.isAbsolute(rel)) {
          return rel;
        }
      }
      return path.basename(normalized);
    }
    return SearchScope.relativizeRgPath(rgFilePath, this.cwd);
  }

  static relativizeRgPath(pathStr: string, searchRoot: string): string {
    const trimmed = pathStr.startsWith('./') ? pathStr.slice(2) : pathStr;
    if (path.isAbsolute(trimmed)) {
      const normalized = path.resolve(trimmed);
      const rel = path.relative(searchRoot, normalized);
      return !rel.startsWith('..') && !path.isAbsolute(rel) ? rel : normalized;
    }
    return path.normalize(trimmed);
  }
}
